modules/jfx.incubator.richtext/src/main/java/jfx/incubator/scene/control/richtext/RichTextArea.java

```diff
@@ -0,0 +1,2345 @@
+ *
+ *   RichTextArea textArea = new RichTextArea();
+ *   // build the content
+ *   textArea.appendText("RichTextArea\n", heading);
+ *   textArea.appendText("Example:\nText is ", StyleAttributeMap.EMPTY);
+ *   textArea.appendText("monospaced.\n", mono);
+ * }</pre>
+ * Which results in the following visual representation:
+ * <p>
+ * <img src="doc-files/RichTextArea.png" alt="Image of the RichTextArea control">
+ * </p>
+ * <p>
+ * A view-only information control requires a different model.  The following example illustrates how to
+ * create a model that uses a style sheet for styling:
+ * <pre>{@code
+ *     SimpleViewOnlyStyledModel m = new SimpleViewOnlyStyledModel();
+ *     // add text segment using CSS style name (requires a style sheet)
+ *     m.addSegment("RichTextArea ", null, "HEADER");
+ *     // add text segment using direct style
+ *     m.addSegment("Demo", "-fx-font-size:200%;", null);
+ *     // newline
+ *     m.nl();
+ *
+ *     RichTextArea textArea = new RichTextArea(m);
+ * }</pre>
+ *
+ * <h2>Text Models</h2>
+ * <p>
+ * A number of standard models can be used with RichTextArea, each addressing a specific use case:
+ * </p>
+ * <table border=1>
+ * <caption>Standard Models</caption>
+ * <tr><th>Model Class</th><th>Description</th></tr>
+ * <tr><td><pre>{@link StyledTextModel}</pre></td><td>Base class (abstract)</td></tr>
+ * <tr><td><pre> ├─ {@link RichTextModel}</pre></td><td>Default model for RichTextArea</td></tr>
+ * <tr><td><pre> ├─ {@link jfx.incubator.scene.control.richtext.model.BasicTextModel BasicTextModel}</pre></td><td>Unstyled text model</td></tr>
+ * <tr><td><pre> │   └─ {@link jfx.incubator.scene.control.richtext.model.CodeTextModel CodeTextModel}</pre></td><td>Default model for CodeArea</td></tr>
+ * <tr><td><pre> └─ {@link jfx.incubator.scene.control.richtext.model.StyledTextModelViewOnlyBase StyledTextModelViewOnlyBase}</pre></td><td>Base class for a view-only model (abstract)</td></tr>
+ * <tr><td><pre>     └─ {@link jfx.incubator.scene.control.richtext.model.SimpleViewOnlyStyledModel SimpleViewOnlyStyledModel}</pre></td><td>In-memory view-only styled model</td></tr>
+ * </table>
+ *
+ * <h2>Selection</h2>
+ * <p>
+ * The RichTextArea control maintains a single {@link #selectionProperty() contiguous selection segment}
+ * as a part of the {@link SelectionModel}.  Additionally,
+ * {@link #anchorPositionProperty()} and {@link #caretPositionProperty()} read-only properties
+ * are derived from the {@link #selectionProperty()} for convenience.
+ *
+ * <h2>Customizing</h2>
+ * The RichTextArea control offers some degree of customization that does not require subclassing:
+ * <ul>
+ * <li>customizing key bindings with the {@link InputMap}
+ * <li>setting {@link #leftDecoratorProperty() leftDecorator}
+ * and {@link #rightDecoratorProperty() rightDecorator} properties
+ * </ul>
+ *
+ * @author Andy Goryachev
+ * @since 999 TODO
+ */
+public class RichTextArea extends Control {
```

Comment:
The *old* `TextField` and `TextArea` controls extend `TextInputControl`.

I wonder whether there is a a _good_ reason not doing the same for `RichTextArea`?
Doing so would allow to more easily to switch between one or the other.

Reply:
Thank you for a very good question!  

The main reason is that RichTextArea (RTA) can work with a large text model which needs to be represented differently from **TextInputControl**'s simple String **text** property.  Other things, like position within the text is also very different, using `{paragraphIndex,charOffset}` instead of a simple `int` offset.  The fact that the model is separated from the control also makes it very different.

At the end, there is simply no benefit in dragging the `TextInputControl` into the picture.

Reply:
I was hoping to be able to easily replace one (old) control with the _new_ one. I mean `RichTextArea` still fulfilling the requirements of `TextInputControl`. Anyhow, I understand you rational not going along this path 👍

Let me provide the use-case I have in mind (which may explain better what I am looking for).
I would like to use/keep textfields and textareas with pure text. Anyhow, having the  possibility to _squiggly_ parts of the text showing spelling mistakes. Since I assume that loading many `RichTextArea` controls (i.e., in tables) will take much longer I expect to offer spelling suggestions (and `RichTextArea`  loading) on request only.

Reply:
It looks like in this particular case you do want to preserve the semantics and APIs of TextArea, rather than switch to a new API.

You could simply extend TextArea/Skin and add functionality to superimpose the squiggly on top of TextArea.  You'll need to listen to several properties and make sure the clipping is set up correctly and you handle the scrolling et cetera, but I think it's doable.

Of course, once you decide that you need to further decorate your text, with let's say highlighted areas, then you might as well use the new component.

By the way, **CodeArea** is specifically designed for that purpose, even has `getText()` and `setText()` - but no textProperty because it's designed to work with large documents.

Reply:
You might be right about impact on performance when trying to use many RichTextArea (RTA) controls as table cells - it is indeed a heavier component, but as long as you ensure there are no memory leaks and your table does not show 1000s of cells on screen we should be ok.

If performance is an issue, a full blown editor might indeed be overkill, and you might be better off created a subclass of Labeled adorned with decorations instead.  You can always set up RTA as an editor if you need to.